File: FEM/2022Autumn/HelpAndOtherVersion/代做python/homework1/hw1.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.tri as tri

mesh = "mesh.txt"
file = open(mesh, 'r')
tmp = file.readlines()
node_num = int(tmp[0])
elem_num = int(tmp[1])
file.close()
xy = np.loadtxt(mesh, dtype=float, skiprows=2, max_rows=node_num)[:, 0:2]
elem = np.loadtxt(mesh, dtype=int, skiprows=2+node_num, max_rows=elem_num)[:, 1:4]
triangle = tri.Triangulation(xy.T[0], xy.T[1], elem)
del tmp

plt.figure(figsize=(10, 10), facecolor="white")
plt.gca().set_aspect(1)
plt.triplot(triangle)
plt.savefig("mesh.png", bbox_inches="tight")
plt.show()

b = 1.0
line = np.sqrt(3) / 3
error = 1e-4
not_boundary_ID = []
for k in range(len(xy)):
    x, y = xy[k]
    error1 = np.abs(x-b)
    error2 = np.abs(y - line * x)
    error3 = np.abs(y + line * x)
    if error1 > error and error2 > error and error3 > error:
        not_boundary_ID.append(k)
        pass
    pass
not_boundary_ID = np.array(not_boundary_ID)

K = np.zeros([node_num, node_num])
B = np.zeros(elem_num)
for k in range(elem_num):
    ID1, ID2, ID3 = elem[k]
    x1, y1 = xy[ID1]
    x2, y2 = xy[ID2]
    x3, y3 = xy[ID3]
    A = 2 * np.linalg.det([
        [1, x1, y1],
        [1, x2, y2],
        [1, x3, y3]
    ])
    k11 = (x2 - x3)**2 + (y2 - y3)**2
    k12 = (x1 - x3) * (-x2 + x3) + (y1 - y3) * (-y2 + y3)
    k13 = (x1 - x2) * (x2 - x3) + (y1 - y2) * (y2 - y3)
    k22 = (x1 - x3)**2 + (y1 - y3)**2
    k23 = -(x1**2 + x2 * x3 - x1 * (x2 + x3) + (y1 - y2) * (y1 - y3))
    k33 = (x1 - x2)**2 + (y1 - y2)**2
    b1 = (x3 * (y1 - y2) + x1 * (y2 - y3) + x2 * (-y1 + y3)) / 6
    b2 = (x3 * (y1 - y2) + x1 * (y2 - y3) + x2 * (-y1+y3)) / 6
    b3 = (x3 * (y1 - y2) + x1 * (y2 - y3) + x2 * (-y1+y3)) / 6

    K[ID1, ID1] += k11 / A;
    K[ID1, ID2] += k12 / A;
    K[ID1, ID3] += k13 / A;

    K[ID2, ID1] += k12 / A;
    K[ID2, ID2] += k22 / A;
    K[ID2, ID3] += k23 / A;

    K[ID3, ID1] += k13 / A;
    K[ID3, ID2] += k23 / A;
    K[ID3, ID3] += k33 / A;

    B[ID1] += b1;
    B[ID2] += b2;
    B[ID3] += b3;
    pass

Km = (K[not_boundary_ID].T[not_boundary_ID]).T
Bm = B[not_boundary_ID]
um = np.linalg.inv(Km) @ Bm
u = np.zeros(node_num)
u[not_boundary_ID] = um

plt.figure(figsize=(10, 10), facecolor="white")
plt.gca().set_aspect(1)
plt.tricontourf(triangle, u, 30)
plt.colorbar()
plt.savefig("result.png", bbox_inches="tight")
plt.show()

File: FEM/2022Autumn/HelpAndOtherVersion/代做matlab/hw1/hw1.m
clear;clc;close all;

NE = textread('mesh.txt', '%d', 2);
N = NE(1);
E = NE(2);
data = readmatrix('mesh.txt');
position = data(1:N, 1:2);
connect = round(data(N+1:N+E, 2:4)) + 1;
b = 10;
alpha = 30*pi/180;

NI = [];
toler = 1e-5;
for i = 1: N
    x = position(i, 1);
    y = position(i, 2);
    a1 = abs(x - b);
    a2 = abs(y - tan(alpha) * x);
    a3 = abs(y + tan(alpha) * x);
    if a1 > toler && a2 > toler && a3 > toler
        NI = [NI, i];
    end
end

K = zeros(N, N);
F = zeros(1, N);
u = zeros(1, N);
for k = 1: E
    node1 = connect(k, 1);
    node2 = connect(k, 2);
    node3 = connect(k, 3);
    x1 = position(node1, 1);
    y1 = position(node1, 2);
    x2 = position(node2, 1);
    y2 = position(node2, 2);
    x3 = position(node3, 1);
    y3 = position(node3, 2);
    A2 = 2*det([1 x1 y1; 1 x2 y2; 1 x3 y3]);
    k11 = (x2 - x3)^2 + (y2 - y3)^2;
    k12 = (x1 - x3) * (-x2 + x3) + (y1 - y3) * (-y2 + y3);
    k13 = (x1 - x2) * (x2 - x3) + (y1 - y2) * (y2 - y3);
    k22 = (x1 - x3)^2 + (y1 - y3)^2;
    k23 = -(x1^2 + x2 * x3 - x1 * (x2 + x3) + (y1 - y2) * (y1 - y3));
    k33 = (x1 - x2)^2 + (y1 - y2)^2;
    F1 = (x3 * (y1 - y2) + x1 * (y2 - y3) + x2 * (-y1 + y3)) / 6;
    F2 = (x3 * (y1 - y2) + x1 * (y2 - y3) + x2 * (-y1+y3)) / 6;
    F3 = (x3 * (y1 - y2) + x1 * (y2 - y3) + x2 * (-y1+y3)) / 6;
    K(node1, node1) = K(node1, node1) + k11 / A2;
    K(node1, node2) = K(node1, node2) + k12 / A2;
    K(node1, node3) = K(node1, node3) + k13 / A2;
    K(node2, node1) = K(node2, node1) + k12 / A2;
    K(node2, node2) = K(node2, node2) + k22 / A2;
    K(node2, node3) = K(node2, node3) + k23 / A2;
    K(node3, node1) = K(node3, node1) + k13 / A2;
    K(node3, node2) = K(node3, node2) + k23 / A2;
    K(node3, node3) = K(node3, node3) + k33 / A2;
    F(node1) = F(node1) + F1;
    F(node2) = F(node2) + F2;
    F(node3) = F(node3) + F3;
end

u(NI) = K(NI, NI) \ F(NI).';
trisurf(connect, position(:, 1), position(:, 2), u)

view([-62.63 75.27])
exportgraphics(gca, 'res1.png')
view([-0.44 90.00])
exportgraphics(gca, 'res2.png')
view([55.41 66.95])
exportgraphics(gca, 'res3.png')
